scrape/manager.go

```diff
@@ -97,71 +108,67 @@ func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	return m.targetsAll
+}
 
-	return targets
+// TargetsActive returns the active targets currently being scraped.
+func (m *Manager) TargetsActive() []*Target {
+	m.mtxTargets.Lock()
+	defer m.mtxTargets.Unlock()
+	return m.targetsActive
 }
 
-// Targets returns the targets currently being scraped.
-func (m *Manager) Targets() []*Target {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
+// TargetsDropped returns the dropped targets during relabelling.
+func (m *Manager) TargetsDropped() []*Target {
+	m.mtxTargets.Lock()
+	defer m.mtxTargets.Unlock()
+	return m.targetsDropped
+}
 
-	var targets []*Target
-	for _, p := range m.scrapePools {
-		p.mtx.RLock()
-		for _, tt := range p.targets {
-			targets = append(targets, tt)
-		}
-		p.mtx.RUnlock()
-	}
+func (m *Manager) targetsUpdate(active, dropped map[string][]*Target) {
+	m.mtxTargets.Lock()
+	defer m.mtxTargets.Unlock()
 
-	return targets
-}
+	m.targetsAll = make(map[string][]*Target)
+	m.targetsActive = nil
+	m.targetsDropped = nil
+	for jobName, targets := range active {
+		m.targetsAll[jobName] = append(m.targetsAll[jobName], targets...)
+		m.targetsActive = append(m.targetsActive, targets...)
 
-// DroppedTargets returns the targets dropped during relabelling.
-func (m *Manager) DroppedTargets() []*Target {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-	var droppedTargets []*Target
-	for _, p := range m.scrapePools {
-		p.mtx.RLock()
-		droppedTargets = append(droppedTargets, p.droppedTargets...)
-		p.mtx.RUnlock()
 	}
-	return droppedTargets
+	for jobName, targets := range dropped {
+		m.targetsAll[jobName] = append(m.targetsAll[jobName], targets...)
+		m.targetsDropped = append(m.targetsDropped, targets...)
+	}
 }
 
+// Should not be called concurrently.
 func (m *Manager) reload(t map[string][]*targetgroup.Group) {
-	for tsetName, tgroup := range t {
-		scrapeConfig, ok := m.scrapeConfigs[tsetName]
-		if !ok {
-			level.Error(m.logger).Log("msg", "error reloading target set", "err", fmt.Sprintf("invalid config id:%v", tsetName))
-			continue
-		}
+	tDropped := make(map[string][]*Target)
+	tActive := make(map[string][]*Target)
 
-		// Scrape pool doesn't exist so start a new one.
-		existing, ok := m.scrapePools[tsetName]
-		if !ok {
-			sp := newScrapePool(scrapeConfig, m.append, log.With(m.logger, "scrape_pool", tsetName))
+	m.mtxScrape.Lock()
+	for tsetName, tgroup := range t {
+		var sp *scrapePool
+		if existing, ok := m.scrapePools[tsetName]; !ok {
+			scrapeConfig, ok := m.scrapeConfigs[tsetName]
+			if !ok {
+				level.Error(m.logger).Log("msg", "error reloading target set", "err", fmt.Sprintf("invalid config id:%v", tsetName))
+				continue
+			}
+			sp = newScrapePool(scrapeConfig, m.append, log.With(m.logger, "scrape_pool", tsetName))
 			m.scrapePools[tsetName] = sp
-			sp.Sync(tgroup)
-
 		} else {
-			existing.Sync(tgroup)
+			sp = existing
 		}
+		tActive[tsetName], tDropped[tsetName] = sp.Sync(tgroup)
 	}
+	m.mtxScrape.Unlock()
+
+	m.targetsUpdate(tActive, tDropped)
```

Comment:
This would be a little cleaner if also protected by mtxScrape. It's not a race as currently written, but the code might change in future.

Reply:
if we add it to the `mtxScrape` than the web UI will also block during `sp.Sync(tgroup)` which is what we are trying to avoid.

Reply:
I mean move this line up to before the m.mtxScrape.Unlock()